external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -240,29 +240,30 @@ private[sql] class AvroSerializer(
       catalystPath: Seq[String],
       avroPath: Seq[String]): InternalRow => Record = {
 
-    val avroPathStr = toFieldStr(avroPath)
-    if (avroStruct.getType != RECORD) {
-      throw new IncompatibleSchemaException(s"$avroPathStr was not a RECORD")
+    val avroSchemaHelper =
+      new AvroUtils.AvroSchemaHelper(avroStruct, catalystStruct, avroPath, positionalFieldMatch)
+
+    avroSchemaHelper.getCatalystFieldsWithoutMatch match {
+      case Seq() => // do nothing
+      case extraFields => throw new IncompatibleSchemaException(
+        s"Cannot find in Avro ${toFieldStr(avroPath)}: " + extraFields
+          .map(f => toFieldStr(catalystPath :+ f.name))
+          .mkString("[", ", ", "]"))
     }
-    val avroFields = avroStruct.getFields.asScala
-    if (avroFields.size != catalystStruct.length) {
-      throw new IncompatibleSchemaException(
-        s"Avro $avroPathStr schema length (${avroFields.size}) doesn't match " +
-        s"SQL ${toFieldStr(catalystPath)} schema length (${catalystStruct.length})")
+    avroSchemaHelper.getAvroFieldsWithoutMatch.toSeq match {
```

Comment:
Same here.

Reply:
Responding on the `AvroDeserializer` comment to keep discussion consolidated.